Watchman.Discord/Areas/Initialization/Services/ServerScanningService.cs

```diff
@@ -20,58 +22,75 @@ public ServerScanningService(ICommandBus commandBus, MessagesHistoryService mess
         {
-            const int LIMIT = 100000;
+            const int LIMIT = 20000; // low limit to use props of startTime optimization
 
             if (channel.Name.Contains("logs"))
                 return;
-
-            var messages = (await _messagesHistoryService.ReadMessagesAsync(server, channel, LIMIT)).ToList();
-            Serilog.Log.Information($"Channel: {channel.Name} read");
-            if (messages.Count == 0)
+            
+            startTime ??= DateTime.UnixEpoch;
+            var messages = ReadMessages(server, channel, limit: 1);
+            if (messages.Count == 0 || LastMessageIsOlderThanStartTime(messages, startTime.Value))
             {
+                Log.Information($"Channel: {channel.Name} has no new messages");
                 return;
             }
 
-            var lastMessageId = messages.Last().Id;
-            await SaveMessages(messages, channel.Id);
-            Serilog.Log.Information($"Channel: {channel.Name} saved");
-
+            var lastMessageId = 0UL;
             do
             {
-                messages = (await _messagesHistoryService.ReadMessagesAsync(server, channel, LIMIT, lastMessageId, goBefore: true)).ToList();
-                Serilog.Log.Information($"Channel: {channel.Name} read");
+                messages = ReadMessages(server, channel, LIMIT, lastMessageId);
                 if (messages.Count == 0)
                 {
                     break;
                 }
 
                 lastMessageId = messages.Last().Id;
                 await SaveMessages(messages, channel.Id);
-                Serilog.Log.Information($"Channel: {channel.Name} saved");
 
-            } while (messages.Count < LIMIT);
+                if (LastMessageIsOlderThanStartTime(messages, startTime.Value))
+                {
+                    break;
+                }
+
+            } while (messages.Count == LIMIT);
+
+            Log.Information($"Channel: {channel.Name} read and saved");
         }
 
-        private async Task SaveMessages(IEnumerable<Devscord.DiscordFramework.Services.Models.Message> messages, ulong channelId)
+        private async Task SaveMessages(IEnumerable<Message> messages, ulong channelId)
         {
             var convertedMessages = ConvertToMessages(messages);
             var command = new AddMessagesCommand(convertedMessages, channelId);
             await _commandBus.ExecuteAsync(command);
         }
 
-        private IEnumerable<Message> ConvertToMessages(IEnumerable<Devscord.DiscordFramework.Services.Models.Message> messages)
+        private IEnumerable<DomainModel.Messages.Message> ConvertToMessages(IEnumerable<Message> messages)
         {
             return messages.Select(x =>
             {
-                var builder = Message.Create(x.Request.OriginalMessage);
+                var builder = DomainModel.Messages.Message.Create(x.Request.OriginalMessage);
                 builder.WithAuthor(x.Contexts.User.Id, x.Contexts.User.Name);
```

Comment:
czemu nie jako 

builder.Create
WithAuthor
.WithChannel
.WithServer 
itd?

chociaż to akurat temat "co kto woli"
w ten sposób można szybciej prześledzić jak obiekt jest budowany
a sposób jaki podałem wyżej jest bardziej estetyczny